### PR TITLE
upstream: implement UpstreamLocalityStats.loadMetricStats

### PR DESCRIPTION
**Pull request by pianiststickman**
Implements a simple API allowing extensions to populate UpstreamLocalityStats.loadMetricStats in load reports, stored as a flat hash map of keys and values that may vary at runtime. Individual stats are accumulated by calling add(), which combines stats with the same name; the aggregated stats are retrieved by calling latch(), which also clears the current load metrics. The load metrics are latched and read by the load stats reporter.

Signed-off-by: Eugene Chan <eugenechan@google.com>

Commit Message: upstream: implement UpstreamLocalityStats.loadMetricStats
Additional Description: Implements a simple API allowing extensions to populate UpstreamLocalityStats.loadMetricStats in load reports, stored as a flat hash map of keys and values that may vary at runtime. Individual stats are accumulated by calling add(), which combines stats with the same name; the aggregated stats are retrieved by calling latch(), which also clears the current load metrics. The load metrics are latched and read by the load stats reporter.

Note: this implementation puts a copy of the stat name into every host that receives a copy of that metric. This can be improved by putting a single copy of the stat name into a thread-local key->index map and using the index as the key to the stat map instead.
Risk Level: low
Testing: CI
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by pianiststickman** [comment]
> I think I'm missing some context as I was not aware of the load-stats-reporter infrastructure.
> 
> Where do these stats get consumed?

These stats are sent as part of LoadStatsRequest to the management server. (See https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/load_report.proto#envoy-v3-api-msg-config-endpoint-v3-upstreamlocalitystats)

File: envoy/upstream/host_description.h
Comment (adisuissa) [inline envoy/upstream/host_description.h:94]: IIUC this adds a per-host map, even in the case where no extension is using `LoadMetricStats`. This adds to the mem usage, which might be significant for use-cases where many endpoints are needed.

File: envoy/upstream/host_description.h
Comment (adisuissa) [inline envoy/upstream/host_description.h:79]: One thing to note is that the name is shared between different extensions (multiple extensions can add to the same key). Is this intended?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Left a few comments with some questions.

File: envoy/upstream/host_description.h
Comment (jmarantz) [reply envoy/upstream/host_description.h]: I feel like I mentioned this in internal comms but can we have a name->index "descriptor map" which is shared by all hosts, and in each host just have a vector of stats accessed via descriptor index?

**Comment by pianiststickman** [comment]
Thanks, PTAL

/retest

File: envoy/upstream/host_description.h
Comment (pianiststickman) [reply envoy/upstream/host_description.h:63]: Done.

File: envoy/upstream/host_description.h
Comment (pianiststickman) [reply envoy/upstream/host_description.h]: @jmarantz I guess the question there would be the potential thread safety issues in updating the descriptor map. Do we then have to lock a resource that's shared across all hosts?

**Review by adisuissa — COMMENTED** [review]
Code changes LGTM modulo the [design comment](https://github.com/envoyproxy/envoy/pull/18534#discussion_r743203959).

File: envoy/upstream/host_description.h
Comment (jmarantz) [reply envoy/upstream/host_description.h]: I think you indicated this before but I had forgotten: the set of names you might want to use for these keys is not known at compile-time, because it comes in headers, so it needs to be dynamically changed? That's the issue?

Maybe you can add comments describing how a string becomes a key into this map.

**Comment by yanavlasov** [comment]
@pianiststickman can you clarify [comments](https://github.com/envoyproxy/envoy/pull/18534#discussion_r743203959) please so we can submit it?

File: envoy/upstream/host_description.h
Comment (jmarantz) [reply envoy/upstream/host_description.h]: Also ISTM that for a given deployment we  could potentially know exactly what the names are, and could lock them in on startup.

Having a string-map for every host with all the strings repeated seems pretty expensive.

Comment (pianiststickman) [reply envoy/upstream/host_description.h]: Correct, the set of names is not known at compile-time. Part of the intent here is to able to utilize this API to implement the Open RCA spec (https://github.com/envoyproxy/envoy/issues/6614) wherein we can arbitrarily define additional metrics. It seems to me that ability runs counter to having the set of keys/metric names pre-defined in a deployment.

Comment (pianiststickman) [reply envoy/upstream/host_description.h]: I expanded the comment to `add()`. It should also be mentioned - the map gets cleared when `latch()` is called. So the keys, if they are present in the map, can be known to be in current use. Once an extension (or the host) stops reporting a metric, that metric will stop taking up memory.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:101]: As discussed I think this of concern for memory use and lock contention. Can you add comments indicating this should be examined before this gets used in scenarios with large numbers of hosts?

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:83]: suggest moving impls to .cc file.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:80]: I think this can be removed.

**Review by jmarantz — COMMENTED** [review]
So this PR has no effect on Envoy other than to make the HostDescription object 8 bytes bigger.

Can you put a pointer to a doc explaining how this can be used, and its scaling limitations, into C++ comments and the PR description?

Thank you!

File: envoy/upstream/host_description.h
Comment (jmarantz) [inline envoy/upstream/host_description.h:81]: nit: absl::string_view on interface.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h:80]: I believe I need to explicitly declare the default constructor since I've deleted the copy and assignment constructors below.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h:87]: Done.

File: envoy/upstream/host_description.h
Comment (pianiststickman) [reply envoy/upstream/host_description.h]: Done.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h:105]: Done.

**Comment by pianiststickman** [comment]
PTAL, and apologies for the delay on this. I've added comments explaining the performance considerations of the current implementation. I think since this is purely an implementation detail (putting key names in thread-local storage, eliminating contention on the locks used to hold the stat maps) I'd prefer to take a pragmatic approach and dispense with that when it becomes necessary.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h:87]: Impls still in .h file?

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:107]: not strictly needed as you have 2 non-copyable members in here, but it's up to you. I suspect that if you do drop these you can also drop the default ctor.

File: envoy/upstream/host_description.h
Comment (jmarantz) [inline envoy/upstream/host_description.h:62]: drop extra line

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:105]: no need to init a unique_ptr; those default-init to nullptr.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:79]: Also note here that it takes a lock on the hot-path.

I really think this should not be deployed as is, but I understand why you want to add the interfaced and iterate on the perf.

File: envoy/upstream/host_description.h
Comment (jmarantz) [inline envoy/upstream/host_description.h:73]: StatMap, since it's a map to a single stat?

File: test/common/upstream/load_stats_reporter_test.cc
Comment (jmarantz) [inline test/common/upstream/load_stats_reporter_test.cc:252]: this test is very long and hard to see the structure of, or what it's trying to do. It's got a lot repetition. Suggest making helper functions in the LoadStatsReporterTest class, or maybe local lambdas that capture the repeated stanzas in one line. For example, you could capture each of the stanzas around here with something like:
```
  void addStats(const HostSharedPtr& host, float a, float b, float c = 0, float d = 0) {
    host->stats().rq_success_.inc();
    host->loadMetricStats().add("metric_a", a);
    host->loadMetricStats().add("metric_b", b);
    if (c != 0) { host->loadMetricStats().add("metric_c", c); }
    ...
}
```
and then these stanzas reduce to one line each:
```
   addStats(host0, 0.11111, 1.0);
   addStats(host0, 0.3333, 0, 3.14159);
   ....
```
Similar factoring can be done in the stanzas below with the expected stats setup. Those helper functions might might make it self-documenting, or we can add some comments.

**Review by jmarantz — COMMENTED** [review]
if you merge main, maybe the coverage test will pass. Otherwise you might have to dive into the CI report and see what's wrong. Sometimes coverage fails because we have added lines of code that are not covered with the new tests, sometimes it's just a flake or a consequence of non-hermetic tests, since the coverage tests run all in one process.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h]: For some reason I misread this as "impls to impl file" originally and got confused.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h]: Done.

File: envoy/upstream/host_description.h
Comment (pianiststickman) [reply envoy/upstream/host_description.h]: Done.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h]: Done.

File: source/common/upstream/upstream_impl.h
Comment (pianiststickman) [reply source/common/upstream/upstream_impl.h]: Done.

File: test/common/upstream/load_stats_reporter_test.cc
Comment (pianiststickman) [reply test/common/upstream/load_stats_reporter_test.cc]: Done, agreed, it looks a lot better now.

File: test/common/upstream/load_stats_reporter_test.cc
Comment (jmarantz) [inline test/common/upstream/load_stats_reporter_test.cc:284]: consider using advanceTimeWait with a delta, rather than just setting the time to an absolute time.

**Review by jmarantz — APPROVED** [review]
/assign-from envoyproxy/senior-maintainers

File: source/common/upstream/upstream_impl.cc
Comment (jmarantz) [inline source/common/upstream/upstream_impl.cc:221]: I just want to call out this warning so the senior-maintainer sees it.

We have more detailed notes on how to avoid taking a lock on every request for this use-case, but @pianiststickman wants to merge this before building out the thread-local impl to establish the API even with a potential perf bottleneck and then iterate.

Note that this code can't be reached in production without adding new C++ APIs, so it will not impact existing Envoy deployments.

**Comment by htuch** [comment]
We need another senior maintainer, since both Josh and PR author are Googlers :)

**Review by zuercher — APPROVED** [review]
Looks fine to me given the proviso that the performance issue addressed before this becomes available for general use.

**Pull request merged**
